src/psyclone/psyGen.py

```diff
@@ -3318,52 +3434,34 @@ def _find_write_arguments(self, nodes, ignore_halos=False):
             return []
 
         # We only need consider nodes that have arguments
-        nodes_with_args = [x for x in nodes if isinstance(x, Call) or
-                           (isinstance(x, HaloExchange) and
-                            not ignore_halos) or
-                           isinstance(x, GlobalSum)]
+        nodes_with_args = [x for x in nodes if
+                           isinstance(x, (Call, GlobalSum)) or
+                           (isinstance(x, HaloExchange) and not ignore_halos)]
+        access = DataAccess(self)
         arguments = []
-        vector_count = 0
         for node in nodes_with_args:
             for argument in node.args:
                 # look at all arguments in our nodes
-                if argument.name != self.name:
-                    # different names so there is no dependence
-                    continue
                 if argument.access not in self._write_access_types:
                     # no dependence if not a writer
                     continue
-                if isinstance(node, HaloExchange):
-                    if isinstance(self.call, HaloExchange):
-                        # no dependence if both nodes are halo
-                        # exchanges (as they act on different vector
-                        # indices).
-                        self.call.check_vector_halos_differ(node)
-                        continue
-                    else:
-                        # a vector read will depend on more than one
-                        # halo exchange (a dependence for each vector
-                        # index) as halo exchanges only act on a
-                        # single vector index
-                        vector_count += 1
-                        arguments.append(argument)
-                        if vector_count == self._vector_size:
-                            # found all of the halo exchange
-                            # dependencies. As they are writers
-                            # we now return
-                            return arguments
-                else:
-                    # this argument is a writer so add it and return
-                    if arguments:
-                        raise GenerationError(
-                            "Internal error, found a writer dependence but "
-                            "there are already dependencies. This should not "
-                            "happen.")
-                    return [argument]
+                if not access.overlaps(argument):
+                    # Accesses are independent of each other
+                    continue
+                arguments.append(argument)
+                access.update_coverage(argument)
+                if access.covered:
```

Comment:
Please add a comment, e.g. "Have found all XXX on which this argument depends so return them" or some such.